zstacklib/zstacklib/utils/gpu.py

```
@@ -665,6 +666,37 @@ def has_nvidia_gpu():
 _nvidia_persistenced_lock = threading.Lock()
 
 
+@lock.lock("gpu-detach-from-host")
+def detach_from_host_with_lock(addr, vendor):
+    pre_detach_from_host(vendor)
+    r, o, e = bash.bash_roe("virsh nodedev-detach pci_%s" % addr.replace(':', '_').replace('.', '_'))
+    if r != 0:
+        raise "failed to nodedev-detach %s: %s, %s" % (addr, o, e)
+
+
+@linux.retry(times=30, sleep_time=5)
```

Ruff (0.14.8)

`linux` may be undefined, or defined from star imports

(F405)